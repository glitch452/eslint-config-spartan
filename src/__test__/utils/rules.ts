import { Linter, Rule } from 'eslint';
import { isOff, isWarn } from '../../utils/index.js';

/**
 * Extract a list of rule names that are 'enabled' i.e. set to 'warn' or 'error'
 * @param rules
 */
export function getEnabledRules(rules: Partial<Record<string, Linter.RuleEntry>> = {}): string[] {
  return Object.entries(rules)
    .filter(([, entry]) => entry && !isOff(entry))
    .map(([name]) => name);
}

/**
 * Extract a list of rule names that are set to 'warn'
 * @param rules
 */
export function getWarnRules(rules: Partial<Record<string, Linter.RuleEntry>> = {}): string[] {
  return Object.entries(rules)
    .filter(([, entry]) => entry && isWarn(entry))
    .map(([name]) => name);
}

/**
 * Extract a list of all the rule names with the given prefix prepended to them
 * @param rules
 * @param prefix
 * @param shouldSort
 */
export function listRules(
  rules:
    | Record<string, Linter.RuleEntry | Rule.RuleModule>
    | Map<string, Linter.RuleEntry | Rule.RuleModule>
    | undefined,
  prefix?: string,
  shouldSort: boolean = false,
): string[] {
  const keys = rules instanceof Map ? [...rules.keys()] : Object.keys(rules ?? {});
  const names = keys.map((x) => `${prefix ? `${prefix}/` : ''}${x}`);
  return shouldSort ? names.sort((a, b) => a.localeCompare(b)) : names;
}

/**
 * Extract a list of rule names that are deprecated and, if provided, prepend the prefix to the rule name
 * @param rules
 * @param prefix
 */
export function getDeprecatedRules(
  rules: Record<string, Rule.RuleModule> | Map<string, Rule.RuleModule> | undefined,
  prefix: string = '',
): string[] {
  const entries = rules instanceof Map ? [...rules.entries()] : Object.entries(rules ?? {});
  return entries.filter(([, rule]) => rule.meta?.deprecated).map(([name]) => `${prefix ? `${prefix}/` : ''}${name}`);
}

interface Docs extends NonNullable<Rule.RuleMetaData['docs']> {
  requiresTypeChecking?: boolean;
}

/**
 * Extract a list of rule names that require type information, if provided, prepend the prefix to the rule name
 * @param rules
 * @param prefix
 */
export function getTypeInfoRequiredRules(
  rules: Record<string, Rule.RuleModule> | Map<string, Rule.RuleModule> | undefined,
  prefix: string = '',
): string[] {
  const entries = rules instanceof Map ? [...rules.entries()] : Object.entries(rules ?? {});
  return (
    entries
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      .filter(([, rule]) => (rule.meta?.docs as Docs).requiresTypeChecking)
      .map(([name]) => `${prefix ? `${prefix}/` : ''}${name}`)
  );
}
