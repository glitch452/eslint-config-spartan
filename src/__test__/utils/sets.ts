type Widen<T> = T | (TSReset.WidenLiteral<T> & {});

/**
 * `A âˆ© B`.
 * Compute the intersection of unique elements contained in both {@link a} and {@link b}.
 * @param a
 * @param b
 * @returns
 */
export function intersection<A, B>(a: Widen<A>[], b: Widen<B>[]): Extract<Widen<A>, Widen<B>>[] {
  const setB = new Set<Widen<A> | Widen<B>>(b);
  return [...new Set<Widen<A>>(a)].filter((x): x is Extract<Widen<A>, Widen<B>> => setB.has(x));
}

/**
 * `A - B` a.k.a. `A / B`.
 * Compute the unique elements in {@link a} that do not also exist in {@link b}.
 * @param a
 * @param b
 * @returns
 */
export function difference<A, B>(a: A[], b: B[]): Widen<A>[] {
  const setB = new Set<A | B>(b);
  return [...new Set<A>(a)].filter((x): x is A => !setB.has(x));
}
