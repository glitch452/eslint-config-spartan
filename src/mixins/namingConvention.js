import { files } from '../utils/index.js';
import { CONFIG_NAME_PREFIX } from '../constants.js';
/** @import { Linter } from 'eslint' */

/**
 * The `namingConvention` mixin adds configuration for the [@typescript-eslint/naming-convention](https://typescript-eslint.io/rules/naming-convention) rule.
 * Note: This mixin requires the `typeEnabled` mixin to be already configured.
 * @param {object} options
 * @param {Linter.Config['files']=} options.files Set the files for this config. By default, this applies to all `.cjs` and `.cjsx` extensions.
 * @param {Linter.Config['files']=} options.testFiles Set the files for the test files variation of this config, which disables the boolean conventions. By default, this applies to all `js/ts` extension variants with `test.ext` and `spec.ext`.
 * @param {boolean=} options.useSnakeCaseVars Enforce snake_case for variable names
 * @param {boolean=} options.enableReactNaming Allow PascalCase for function names to accommodate react components
 * @param {boolean | 'non-strict'=} options.enableStoryNaming Allow StrictPascalCase for top-level constant names to accommodate storybook stories
 * @param {boolean=} options.allowUpperCaseVars Allow UPPER_CASE for variable names
 * @param {boolean=} options.allowUpperCaseFuncNames Allow UPPER_CASE for function names
 * @param {Array<Record<string, unknown>>=} options.extraConfigs Extra configuration blocks to append to the generated non-test-file config
 * @param {Array<Record<string, unknown>>=} options.extraTestFileConfigs Extra configuration blocks to append to the generated test file config
 * @param {('forbid' | 'require' | 'requireDouble' | 'allow' | 'allowDouble' | 'allowSingleOrDouble')=} options.privateUnderscore Enforce the presence or absence of a leading underscore for private class members
 * @param {Linter.Config['ignores']=} options.ignores Set the ignores for this config. By default, this applies to all `.md/*` and `.mdx/*` to ignore code within markdown file code blocks.
 * @returns {Linter.Config[]}
 */
export function namingConvention(options = {}) {
  return [
    {
      name: `${CONFIG_NAME_PREFIX}/${namingConvention.name}`,
      files: options.files ?? [files.jsTs],
      ignores: options.ignores ?? [`${files.mdMdx}/**/*`],
      rules: {
        '@typescript-eslint/naming-convention': [
          'error',
          ...getBaseComponents(),
          ...getPrivateMembersComponents(options),
          ...getFunctionComponents(options),
          ...getVariableNameComponents(options),
          ...getBooleanComponents(options),
          ...(options.extraConfigs ?? []),
        ],
      },
    },
    {
      name: `${CONFIG_NAME_PREFIX}/${namingConvention.name}TestFiles`,
      files: options.testFiles ?? [files.testSpec],
      ignores: options.ignores ?? [`${files.mdMdx}/**/*`],
      rules: {
        '@typescript-eslint/naming-convention': [
          'error',
          ...getBaseComponents(),
          ...getPrivateMembersComponents(options),
          ...getFunctionComponents(options),
          ...getVariableNameComponents(options),
          ...(options.extraTestFileConfigs ?? []),
        ],
      },
    },
  ];
}

function getBaseComponents() {
  return [
    // Enum members must be UPPER_CASE since they are treated as constants
    { selector: ['enumMember'], format: ['UPPER_CASE'] },

    // Allow destructured variables to keep their name formatting
    {
      selector: ['variable', 'parameter'],
      modifiers: ['destructured'],
      format: null,
      filter: { regex: '.+', match: true }, // Required to take precedence over other rules
    },
    {
      selector: ['variable', 'parameter'],
      modifiers: ['destructured'],
      format: null,
      types: ['function'], // Required to take precedence over other rules
      filter: { regex: '.+', match: true }, // Required to take precedence over other rules
    },

    // `typeLike` (class, interface, typeAlias, enum, typeParameter) should use StrictPascalCase
    { selector: 'typeLike', format: ['StrictPascalCase'] },
  ];
}

/**
 * @param {object} options
 * @param {boolean=} options.useSnakeCaseVars Enforce snake_case for variable names
 * @param {boolean=} options.allowUpperCaseVars Allow UPPER_CASE for variable names
 * @param {('forbid' | 'require' | 'requireDouble' | 'allow' | 'allowDouble' | 'allowSingleOrDouble')=} options.privateUnderscore Enforce the presence or absence of a leading underscore for private class members
 */
function getPrivateMembersComponents({ useSnakeCaseVars, privateUnderscore, allowUpperCaseVars }) {
  const snakeCaseVars = ['snake_case', 'UPPER_CASE'];
  const nonSnakeCaseVars = ['strictCamelCase'];
  if (allowUpperCaseVars) {
    nonSnakeCaseVars.push('UPPER_CASE');
  }

  return !privateUnderscore
    ? []
    : [
        {
          selector: ['accessor', 'classProperty'],
          format: useSnakeCaseVars ? snakeCaseVars : nonSnakeCaseVars,
          modifiers: ['private'],
          leadingUnderscore: privateUnderscore,
        },
      ];
}

/**
 * @param {object} options
 * @param {boolean=} options.useSnakeCaseVars Enforce snake_case for variable names
 * @param {boolean=} options.enableReactNaming Allow PascalCase for function names to accommodate react components
 * @param {boolean=} options.allowUpperCaseFuncNames Allow UPPER_CASE for function names
 */
function getFunctionComponents({ useSnakeCaseVars, enableReactNaming, allowUpperCaseFuncNames }) {
  const reactNames = ['strictCamelCase', 'StrictPascalCase'];
  const snakeCaseNames = ['snake_case', 'UPPER_CASE'];
  const nonSnakeCaseNames = ['strictCamelCase'];

  if (allowUpperCaseFuncNames) {
    reactNames.push('UPPER_CASE');
    nonSnakeCaseNames.push('UPPER_CASE');
  }

  return [
    // Functions must use strictCamelCase
    {
      selector: ['function', 'parameter', 'variable', 'classMethod', 'typeMethod', 'classProperty', 'typeProperty'],
      format: enableReactNaming ? reactNames : nonSnakeCaseNames,
      types: ['function'], // To scope 'parameter', 'classProperty', 'typeProperty' and 'variable'
      filter: { regex: 'toJSON', match: false },
    },
    // Unused function parameters must begin with an underscore
    {
      selector: ['parameter'],
      format: useSnakeCaseVars ? snakeCaseNames : nonSnakeCaseNames,
      modifiers: ['unused'],
      leadingUnderscore: 'require',
    },
    {
      selector: ['parameter'],
      format: ['strictCamelCase'],
      modifiers: ['unused'],
      types: ['function'],
      filter: { regex: '.+', match: true }, // Required to take precedence over other rules
      leadingUnderscore: 'require',
    },
  ];
}

/**
 * @param {object} options
 * @param {boolean=} options.useSnakeCaseVars Enforce snake_case for variable names
 * @param {boolean=} options.allowUpperCaseVars Allow UPPER_CASE for variable names
 * @param {boolean | 'non-strict'=} options.enableStoryNaming Allow StrictPascalCase for top-level constant names to accommodate storybook stories
 */
function getVariableNameComponents({ useSnakeCaseVars, allowUpperCaseVars, enableStoryNaming }) {
  const snakeCaseVars = ['snake_case', 'UPPER_CASE'];
  const nonSnakeCaseVars = ['strictCamelCase'];
  const topLevelConstants = useSnakeCaseVars ? snakeCaseVars : ['UPPER_CASE', ...nonSnakeCaseVars];

  if (allowUpperCaseVars) {
    nonSnakeCaseVars.push('UPPER_CASE');
  }

  if (enableStoryNaming) {
    topLevelConstants.push(enableStoryNaming === 'non-strict' ? 'PascalCase' : 'StrictPascalCase');
  }

  return [
    // Variables in general
    {
      selector: ['variable', 'parameter', 'accessor', 'classProperty', 'typeProperty'],
      format: useSnakeCaseVars ? snakeCaseVars : nonSnakeCaseVars,
      leadingUnderscore: 'allow',
    },
    // Allow an exception for names that are quoted (i.e. 'Content-Type')
    {
      selector: ['objectLiteralProperty'],
      format: null,
      modifiers: ['requiresQuotes'],
    },
    // Always allow UPPER_CASE for top-level constants
    {
      selector: ['variable'],
      format: topLevelConstants,
      modifiers: ['const', 'global'],
    },
  ];
}

/**
 * @param {object} options
 * @param {boolean=} options.useSnakeCaseVars Enforce snake_case for variable names
 * @param {boolean=} options.allowUpperCaseVars Allow UPPER_CASE for variable names
 */
function getBooleanComponents({ useSnakeCaseVars, allowUpperCaseVars }) {
  const nonSnakeCaseVars = ['strictCamelCase'];
  if (allowUpperCaseVars) {
    nonSnakeCaseVars.push('UPPER_CASE');
  }

  const indicators = ['is', 'are', 'was', 'should', 'has', 'can', 'did', 'will', 'allow', 'use', 'requires', 'does'];
  const indicatorsUpperFirst = indicators.map((x) => `${x[0].toUpperCase()}${x.slice(1)}`);

  const lowerSnakePattern = `.*(${indicators.map((x) => `${x}_`).join('|')}).*`;
  const upperSnakePattern = lowerSnakePattern.toUpperCase();
  const camelCasePattern = `(^(${indicators.join('|')})([A-Z].*|$)|.*(${indicatorsUpperFirst.join('|')}).*)`;

  /** @type {Record<string, unknown>[]} */
  const entries = [
    // Handle top-level UPPER_CASE booleans
    {
      selector: ['variable'],
      format: ['UPPER_CASE'],
      modifiers: ['const', 'global'],
      types: ['boolean'],
      custom: { regex: upperSnakePattern, match: true },
      filter: { regex: '.+', match: true }, // Required to take precedence over other rules
    },
  ];

  // Booleans should have an appropriate indicator
  if (useSnakeCaseVars) {
    entries.push(
      // Handle unused parameter booleans
      {
        selector: ['parameter'],
        format: ['snake_case'],
        modifiers: ['unused'],
        types: ['boolean'],
        custom: { regex: lowerSnakePattern, match: true },
        filter: { regex: '^_?[a-z]', match: true }, // Only test snake_case in this rule to avoid is_MIXED
        leadingUnderscore: 'require',
      },
      {
        selector: ['parameter'],
        format: ['UPPER_CASE'],
        modifiers: ['unused'],
        types: ['boolean'],
        custom: { regex: upperSnakePattern, match: true },
        filter: { regex: '^_?[A-Z]', match: true }, // Only test UPPER_CASE in this rule to avoid IS_mixed
        leadingUnderscore: 'require',
      },
      // Handle general case booleans
      {
        selector: ['variable', 'parameter', 'classProperty', 'typeProperty'],
        format: ['snake_case'],
        types: ['boolean'],
        custom: { regex: lowerSnakePattern, match: true },
        filter: { regex: '^[a-z]', match: true }, // Only test snake_case in this rule to avoid is_MIXED
      },
      {
        selector: ['variable', 'parameter', 'classProperty', 'typeProperty'],
        format: ['UPPER_CASE'],
        types: ['boolean'],
        custom: { regex: upperSnakePattern, match: true },
        filter: { regex: '^[A-Z]', match: true }, // Only test UPPER_CASE in this rule to avoid IS_mixed
      },
    );
  } else {
    entries.push(
      // Handle unused parameter booleans
      {
        selector: ['parameter'],
        format: nonSnakeCaseVars,
        modifiers: ['unused'],
        types: ['boolean'],
        custom: { regex: camelCasePattern, match: true },
        leadingUnderscore: 'require',
      },
      // Handle general case booleans
      {
        selector: ['variable', 'parameter', 'classProperty', 'typeProperty'],
        format: nonSnakeCaseVars,
        types: ['boolean'],
        custom: { regex: camelCasePattern, match: true },
      },
    );
  }

  return entries;
}
