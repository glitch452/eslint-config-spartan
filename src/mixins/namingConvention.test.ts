import typescriptEsLint from 'typescript-eslint';
import { namingConvention } from './namingConvention.js';
import { configSchema } from '../__test__/utils/configSchema.js';
import { difference, intersection } from '../__test__/utils/sets.js';
import { files } from '../utils/index.js';
import { CONFIG_NAME_PREFIX, prefixes } from '../constants.js';
import { getDeprecatedRules, getEnabledRules, getWarnRules, listRules } from '../__test__/utils/rules.js';

describe(namingConvention.name, () => {
  const configs = namingConvention();

  const testTable = [
    {
      id: namingConvention.name,
      filesOption: 'files',
      expectedFiles: [files.jsTs],
      extraConfigsOption: [{ test: 'extraConfigsOption' }],
      extraTestFileConfigs: [{ test: 'extraTestFileConfigs' }],
      expectedExtraConfigs: [{ test: 'extraConfigsOption' }],
    },
    {
      id: `${namingConvention.name}TestFiles`,
      filesOption: 'testFiles',
      expectedFiles: [files.testSpec],
      extraConfigsOption: [{ test: 'extraConfigsOption' }],
      extraTestFileConfigs: [{ test: 'extraTestFileConfigs' }],
      expectedExtraConfigs: [{ test: 'extraTestFileConfigs' }],
    },
  ] as const;

  const validRules = listRules((typescriptEsLint.plugin as any).rules, prefixes.typescriptEsLint);
  const deprecatedRules = getDeprecatedRules((typescriptEsLint.plugin as any).rules, prefixes.typescriptEsLint);

  describe.each(testTable)('$id', (item) => {
    const { id, expectedFiles, filesOption, extraConfigsOption, extraTestFileConfigs, expectedExtraConfigs } = item;
    const index = testTable.indexOf(item);
    const config = configs[index];

    it('should create a valid eslint config', () => {
      const actual = () => configSchema.parse(config);
      expect(actual).not.toThrow();
    });

    it('should produce the expected default configuration', () => {
      expect(config.rules).toMatchSnapshot();
    });

    it('should set the name value', () => {
      expect(config.name).toBe(`${CONFIG_NAME_PREFIX}/${id}`);
    });

    it('should set the default files value', () => {
      expect(config.files).toStrictEqual(expectedFiles);
    });

    it('should set the given files value', () => {
      const fileValue = ['files.txt'];
      const actual = namingConvention({ [filesOption]: fileValue })[index].files;
      expect(actual).toStrictEqual(fileValue);
    });

    it('should add the leading underscore config when the "privateUnderscore" option is set', () => {
      const actual = namingConvention({ privateUnderscore: 'allowSingleOrDouble' })[index];
      const expected = expect.arrayContaining([
        expect.objectContaining({
          selector: ['accessor', 'classProperty'],
          modifiers: ['private'],
          leadingUnderscore: 'allowSingleOrDouble',
        }),
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should not add the leading underscore config when the "privateUnderscore" option is not set', () => {
      const actual = namingConvention()[index];
      const expected = expect.arrayContaining([
        expect.objectContaining({
          selector: ['accessor', 'classProperty'],
          modifiers: ['private'],
          leadingUnderscore: 'allowSingleOrDouble',
        }),
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).not.toStrictEqual(expected);
    });

    it('should add the leading underscore config with snake case when the "privateUnderscore" and "useSnakeCaseVars" options are set', () => {
      const actual = namingConvention({ privateUnderscore: 'allowSingleOrDouble', useSnakeCaseVars: true })[index];
      const expected = expect.arrayContaining([
        expect.objectContaining({
          selector: ['accessor', 'classProperty'],
          format: ['snake_case', 'UPPER_CASE'],
          modifiers: ['private'],
          leadingUnderscore: 'allowSingleOrDouble',
        }),
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should allow StrictPascalCase for function names when the "react" option is set to true', () => {
      const actual = namingConvention({ enableReactNaming: true })[index];
      const expected = expect.arrayContaining([
        expect.objectContaining({
          selector: ['function', 'parameter', 'variable', 'classMethod', 'typeMethod', 'classProperty', 'typeProperty'],
          format: ['strictCamelCase', 'StrictPascalCase'],
          types: ['function'],
        }),
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should not allow StrictPascalCase for function names when the "react" option is set to false', () => {
      const actual = namingConvention({ enableReactNaming: false })[index];
      const expected = expect.arrayContaining([
        expect.objectContaining({
          selector: ['function', 'parameter', 'variable', 'classMethod', 'typeMethod', 'classProperty', 'typeProperty'],
          format: ['strictCamelCase'],
          types: ['function'],
        }),
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should enforce snake_case function parameter names when the "useSnakeCaseVars" option is set to true', () => {
      const actual = namingConvention({ useSnakeCaseVars: true })[index];
      const expected = expect.arrayContaining([
        {
          selector: ['parameter'],
          format: ['snake_case', 'UPPER_CASE'],
          modifiers: ['unused'],
          leadingUnderscore: 'require',
        },
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should enforce strictCamelCase function parameter names when the "useSnakeCaseVars" option is set to false', () => {
      const actual = namingConvention({ useSnakeCaseVars: false })[index];
      const expected = expect.arrayContaining([
        { selector: ['parameter'], format: ['strictCamelCase'], modifiers: ['unused'], leadingUnderscore: 'require' },
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should enforce snake_case variable names when the "useSnakeCaseVars" option is set to true', () => {
      const actual = namingConvention({ useSnakeCaseVars: true })[index];
      const expected = expect.arrayContaining([
        {
          selector: ['variable', 'parameter', 'accessor', 'classProperty', 'typeProperty'],
          format: ['snake_case', 'UPPER_CASE'],
          leadingUnderscore: 'allow',
        },
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should enforce strictCamelCase variable names when the "useSnakeCaseVars" option is set to false', () => {
      const actual = namingConvention({ useSnakeCaseVars: false })[index];
      const expected = expect.arrayContaining([
        {
          selector: ['variable', 'parameter', 'accessor', 'classProperty', 'typeProperty'],
          format: ['strictCamelCase'],
          leadingUnderscore: 'allow',
        },
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should allow UPPER_CASE variable names when the "allowUpperCaseVars" option is set to true', () => {
      const nonSnakeCaseVars = ['strictCamelCase', 'UPPER_CASE'];
      const actual = namingConvention({ allowUpperCaseVars: true, privateUnderscore: 'allow' })[index];
      const components: any[] = [
        {
          selector: ['accessor', 'classProperty'],
          format: nonSnakeCaseVars,
          modifiers: ['private'],
          leadingUnderscore: 'allow',
        },
        {
          selector: ['variable', 'parameter', 'accessor', 'classProperty', 'typeProperty'],
          format: nonSnakeCaseVars,
          leadingUnderscore: 'allow',
        },
      ];
      if (index === 0) {
        components.push(
          {
            selector: ['parameter'],
            format: nonSnakeCaseVars,
            modifiers: ['unused'],
            types: ['boolean'],
            // eslint-disable-next-line vitest/no-conditional-expect
            custom: { regex: expect.any(String), match: true },
            leadingUnderscore: 'require',
          },
          {
            selector: ['variable', 'parameter', 'classProperty', 'typeProperty'],
            format: nonSnakeCaseVars,
            types: ['boolean'],
            // eslint-disable-next-line vitest/no-conditional-expect
            custom: { regex: expect.any(String), match: true },
          },
        );
      }
      const expected = expect.arrayContaining(components);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should allow UPPER_CASE function names when the "allowUpperCaseFuncNames" option is set to true', () => {
      const nonSnakeCaseNames = ['strictCamelCase', 'UPPER_CASE'];
      const actual = namingConvention({ allowUpperCaseFuncNames: true })[index];
      const expected = expect.arrayContaining([
        {
          selector: ['function', 'parameter', 'variable', 'classMethod', 'typeMethod', 'classProperty', 'typeProperty'],
          format: nonSnakeCaseNames,
          types: ['function'],
          filter: { regex: 'toJSON', match: false },
        },
        {
          selector: ['parameter'],
          format: nonSnakeCaseNames,
          modifiers: ['unused'],
          leadingUnderscore: 'require',
        },
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should allow UPPER_CASE function names when the "allowUpperCaseFuncNames" and "enableReactNaming" options are set to true', () => {
      const reactNames = ['strictCamelCase', 'StrictPascalCase', 'UPPER_CASE'];
      const actual = namingConvention({ allowUpperCaseFuncNames: true, enableReactNaming: true })[index];
      const expected = expect.arrayContaining([
        {
          selector: ['function', 'parameter', 'variable', 'classMethod', 'typeMethod', 'classProperty', 'typeProperty'],
          format: reactNames,
          types: ['function'],
          filter: { regex: 'toJSON', match: false },
        },
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should allow StrictPascalCase top-level constant names when the "enableStoryNaming" option is set to true', () => {
      const actual = namingConvention({ enableStoryNaming: true })[index];
      const expected = expect.arrayContaining([
        {
          selector: ['variable'],
          format: ['UPPER_CASE', 'strictCamelCase', 'StrictPascalCase'],
          modifiers: ['const', 'global'],
        },
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should allow PascalCase top-level constant names when the "enableStoryNaming" option is set to "non-strict"', () => {
      const actual = namingConvention({ enableStoryNaming: 'non-strict' })[index];
      const expected = expect.arrayContaining([
        {
          selector: ['variable'],
          format: ['UPPER_CASE', 'strictCamelCase', 'PascalCase'],
          modifiers: ['const', 'global'],
        },
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should allow StrictPascalCase top-level constant names when the "enableStoryNaming" and "useSnakeCaseVars" options are set to true', () => {
      const actual = namingConvention({ enableStoryNaming: true, useSnakeCaseVars: true })[index];
      const expected = expect.arrayContaining([
        {
          selector: ['variable'],
          format: ['snake_case', 'UPPER_CASE', 'StrictPascalCase'],
          modifiers: ['const', 'global'],
        },
      ]);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should append the values provided via the "extraConfigs" option', () => {
      const actual = namingConvention({
        extraConfigs: extraConfigsOption as any,
        extraTestFileConfigs: extraTestFileConfigs as any,
      })[index];
      const expected = expect.arrayContaining(expectedExtraConfigs as any);
      expect(actual.rules!['@typescript-eslint/naming-convention']).toStrictEqual(expected);
    });

    it('should only configure rules that exist', () => {
      const configuredRules = Object.keys(config.rules ?? {});
      const actual = difference(configuredRules, validRules);
      expect(actual).toStrictEqual([]);
    });

    it('should not configure rules that are deprecated', () => {
      const configuredAndEnabledRules = getEnabledRules(config.rules);
      const actual = intersection(configuredAndEnabledRules, deprecatedRules);
      expect(actual).toStrictEqual([]);
    });

    it('should not configure any rules as warn', () => {
      const configuredAndWarnRules = getWarnRules(config.rules);
      expect(configuredAndWarnRules).toStrictEqual([]);
    });
  });
});
