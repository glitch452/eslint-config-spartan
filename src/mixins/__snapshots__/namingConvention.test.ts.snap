// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`namingConvention > 'namingConvention' > should produce the expected default configuration 1`] = `
{
  "@typescript-eslint/naming-convention": [
    "error",
    {
      "format": [
        "UPPER_CASE",
      ],
      "selector": [
        "enumMember",
      ],
    },
    {
      "filter": {
        "match": true,
        "regex": ".+",
      },
      "format": null,
      "modifiers": [
        "destructured",
      ],
      "selector": [
        "variable",
        "parameter",
      ],
    },
    {
      "filter": {
        "match": true,
        "regex": ".+",
      },
      "format": null,
      "modifiers": [
        "destructured",
      ],
      "selector": [
        "variable",
        "parameter",
      ],
      "types": [
        "function",
      ],
    },
    {
      "format": [
        "StrictPascalCase",
      ],
      "selector": "typeLike",
    },
    {
      "filter": {
        "match": false,
        "regex": "toJSON",
      },
      "format": [
        "strictCamelCase",
      ],
      "selector": [
        "function",
        "parameter",
        "variable",
        "classMethod",
        "typeMethod",
        "classProperty",
        "typeProperty",
      ],
      "types": [
        "function",
      ],
    },
    {
      "format": [
        "strictCamelCase",
      ],
      "leadingUnderscore": "require",
      "modifiers": [
        "unused",
      ],
      "selector": [
        "parameter",
      ],
    },
    {
      "filter": {
        "match": true,
        "regex": ".+",
      },
      "format": [
        "strictCamelCase",
      ],
      "leadingUnderscore": "require",
      "modifiers": [
        "unused",
      ],
      "selector": [
        "parameter",
      ],
      "types": [
        "function",
      ],
    },
    {
      "format": [
        "strictCamelCase",
      ],
      "leadingUnderscore": "allow",
      "selector": [
        "variable",
        "parameter",
        "accessor",
        "classProperty",
        "typeProperty",
      ],
    },
    {
      "format": null,
      "modifiers": [
        "requiresQuotes",
      ],
      "selector": [
        "objectLiteralProperty",
      ],
    },
    {
      "format": [
        "UPPER_CASE",
        "strictCamelCase",
      ],
      "modifiers": [
        "const",
        "global",
      ],
      "selector": [
        "variable",
      ],
    },
    {
      "custom": {
        "match": true,
        "regex": ".*(IS_|ARE_|WAS_|SHOULD_|HAS_|CAN_|DID_|WILL_|ALLOW_|USE_|REQUIRES_).*",
      },
      "filter": {
        "match": true,
        "regex": ".+",
      },
      "format": [
        "UPPER_CASE",
      ],
      "modifiers": [
        "const",
        "global",
      ],
      "selector": [
        "variable",
      ],
      "types": [
        "boolean",
      ],
    },
    {
      "custom": {
        "match": true,
        "regex": "(^(is|are|was|should|has|can|did|will|allow|use|requires)([A-Z].*|$)|.*(Is|Are|Was|Should|Has|Can|Did|Will|Allow|Use|Requires).*)",
      },
      "format": [
        "strictCamelCase",
      ],
      "leadingUnderscore": "require",
      "modifiers": [
        "unused",
      ],
      "selector": [
        "parameter",
      ],
      "types": [
        "boolean",
      ],
    },
    {
      "custom": {
        "match": true,
        "regex": "(^(is|are|was|should|has|can|did|will|allow|use|requires)([A-Z].*|$)|.*(Is|Are|Was|Should|Has|Can|Did|Will|Allow|Use|Requires).*)",
      },
      "format": [
        "strictCamelCase",
      ],
      "selector": [
        "variable",
        "parameter",
        "classProperty",
        "typeProperty",
      ],
      "types": [
        "boolean",
      ],
    },
  ],
}
`;

exports[`namingConvention > 'namingConventionTestFiles' > should produce the expected default configuration 1`] = `
{
  "@typescript-eslint/naming-convention": [
    "error",
    {
      "format": [
        "UPPER_CASE",
      ],
      "selector": [
        "enumMember",
      ],
    },
    {
      "filter": {
        "match": true,
        "regex": ".+",
      },
      "format": null,
      "modifiers": [
        "destructured",
      ],
      "selector": [
        "variable",
        "parameter",
      ],
    },
    {
      "filter": {
        "match": true,
        "regex": ".+",
      },
      "format": null,
      "modifiers": [
        "destructured",
      ],
      "selector": [
        "variable",
        "parameter",
      ],
      "types": [
        "function",
      ],
    },
    {
      "format": [
        "StrictPascalCase",
      ],
      "selector": "typeLike",
    },
    {
      "filter": {
        "match": false,
        "regex": "toJSON",
      },
      "format": [
        "strictCamelCase",
      ],
      "selector": [
        "function",
        "parameter",
        "variable",
        "classMethod",
        "typeMethod",
        "classProperty",
        "typeProperty",
      ],
      "types": [
        "function",
      ],
    },
    {
      "format": [
        "strictCamelCase",
      ],
      "leadingUnderscore": "require",
      "modifiers": [
        "unused",
      ],
      "selector": [
        "parameter",
      ],
    },
    {
      "filter": {
        "match": true,
        "regex": ".+",
      },
      "format": [
        "strictCamelCase",
      ],
      "leadingUnderscore": "require",
      "modifiers": [
        "unused",
      ],
      "selector": [
        "parameter",
      ],
      "types": [
        "function",
      ],
    },
    {
      "format": [
        "strictCamelCase",
      ],
      "leadingUnderscore": "allow",
      "selector": [
        "variable",
        "parameter",
        "accessor",
        "classProperty",
        "typeProperty",
      ],
    },
    {
      "format": null,
      "modifiers": [
        "requiresQuotes",
      ],
      "selector": [
        "objectLiteralProperty",
      ],
    },
    {
      "format": [
        "UPPER_CASE",
        "strictCamelCase",
      ],
      "modifiers": [
        "const",
        "global",
      ],
      "selector": [
        "variable",
      ],
    },
  ],
}
`;
